SwiftCompilerSources/Sources/Optimizer/FunctionPasses/LoopInvariantCodeMotion.swift

```diff
@@ -0,0 +1,1020 @@
+         return singleValueInst != otherInst && singleValueInst.isIdenticalTo(otherInst)
+    }) {
+      guard let identicalSingleValueInst = identicalInst as? SingleValueInstruction else {
+        return true
+      }
+      
+      singleValueInst.replace(with: identicalSingleValueInst, context)
+    }
+
+    return true
+  }
+  
+  func sink(outOf loop: Loop, _ context: FunctionPassContext) -> Bool {
+    var canMove = loop.hasSingleExitBlock
+    let exitBlocks = loop.exitBlocks
+    let exitingBlocks = loop.exitingBlocks
+    var newExitBlocks = Stack<BasicBlock>(context)
+    defer { newExitBlocks.deinitialize() }
+    var changed = false
+
+    for exitingBlock in exitingBlocks {
+      for succesor in exitingBlock.successors where !newExitBlocks.contains(succesor) && exitBlocks.contains(succesor) {
+
+        newExitBlocks.push(succesor)
+
+        if succesor.instructions.contains(where: { isIdenticalTo($0) }) {
+          canMove = false
+        } else if canMove {
+          move(before: succesor.instructions.first!, context)
+        } else if let firstInstruction = succesor.instructions.first {
+          copy(before: firstInstruction, context)
+        } else {
+          continue
+        }
+
+        changed = true
+      }
+    }
+
+    if changed && !canMove {
+      context.erase(instruction: self)
+    }
+
+    return changed
+  }
+  
+  /// Returns `true` if `sideEffect` cannot be reordered with a call to this
+  /// global initializer.
+  private func globalInitMayConflictWith(
+    sideEffect: Instruction,
+    _ aliasAnalysis: AliasAnalysis
+  ) -> Bool {
+    switch sideEffect {
+    case let storeInst as StoreInst:
+      return mayReadOrWrite(address: storeInst.destinationOperand.value, aliasAnalysis)
+    case let loadInst as LoadInst:
+      return mayWrite(toAddress: loadInst.operand.value, aliasAnalysis)
+    case is CondFailInst:
+      return false
+    default:
+      return true
+    }
+  }
+  
+  /// Returns `true` if any of the instructions in `sideEffects` cannot be
+  /// reordered with a call to this global initializer (which is in the same basic
+  /// block).
+  func globalInitMayConflictWith(
+    blockSideEffectSegment: StackWithCount<Instruction>.Segment,
+    _ aliasAnalysis: AliasAnalysis
+  ) -> Bool {
+    return blockSideEffectSegment
+      .contains { sideEffect in
+        globalInitMayConflictWith(
+          sideEffect: sideEffect,
+          aliasAnalysis
+        )
+      }
+  }
+
+  /// Returns `true` if any of the instructions in `loopSideEffects` which are
+  /// post-dominated by a call to this global initializer cannot be reordered with
+  /// the call.
+  func globalInitMayConflictWith(
+    loopSideEffects: StackWithCount<Instruction>,
+    notPostDominatingPreheader: BasicBlock,
+    _ aliasAnalysis: AliasAnalysis,
+    _ postDomTree: PostDominatorTree
+  ) -> Bool {
+    guard parentBlock.postDominates(notPostDominatingPreheader, postDomTree) else {
```

Comment:
It is better to move this check to the call site, because it has nothing to do with checking conflicts